### PR TITLE
Adding genes forward merge process to production

### PR DESCRIPTION
**Pull request by afg1**
The current implementation of the genes workflow in the pipeline misses the part where we merge the previous and new gene sets. This should add the required merging and handling of uploading new genes as well as marking discarded genes inactive.

I've already made the required changes in the database to facilitate this, and we will need to add some flags to the local.config to tell the pipeline what release version we are working with. 

The code has been linted, but has not yet been tested because we need precompute completed and sequence feature views updated for this to work.

### CODE REVIEW

File: genes.nf [hunk 1]
```diff
@@ -1,99 +1,287 @@
...
+  | set { merged_existing }
+
+  // New taxon: nothing to merge, just initialise version tracking.
+  previous_genes.new_taxon \
+  | map { taxid, this_release, _previous, _prev_release -> [taxid, this_release] } \
+  | init_genes \
+  | set { merged_new }
+
+  // Store, then deactivate genes the merge dropped, then metadata.
+  merged_existing.mix( merged_new ) \
```
[the remark this answers is not in the document]
Comment (afg1) [reply genes.nf:274]: Leaving as-is for now. This pipeline runs on SLURM with Singularity and a shared filesystem (`config/cluster.config`), where task work directories are reachable from the head node, so the `.text` reads work in our deployment. The suggested combine of `forward_merge`/`init_genes` is a larger refactor we're deferring; noting it here for future portability to non-shared-FS executors.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: genes.nf [hunk 2]
```diff
@@ -1,99 +1,287 @@
...
-  load_data(locus, rejected, load, load_status, post_load)
+  // _flag is an ordering trigger so this stage can be chained after upstream
+  // stages in main.nf. It is not otherwise used: the pipeline is driven by
+  // params.genes. Standalone runs pass Channel.of(true) (see entry below).
+  take: _flag
+
+  main:
+  Channel.of(true) | fetch_so_model | set { so_model }
+  Channel.of(true) | fetch_rf_model | set { rf_model }
+  Channel.of(true) | fetch_taxids | set { taxa }
```
[the remark this answers is not in the document]
Comment (afg1) [reply genes.nf]: Deferring this one. `_flag` is intentionally an ordering placeholder (see the workflow comment); wiring it through to the three entry processes needs care around Nextflow's consume-once channel semantics, so we're handling it separately rather than in this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Vicbeg** [comment]
Hi Andrew, I think your code looks good, but I asked codex to review with the wider context in mind and it found these potential bugs. I'll paste here:

Findings
**High, Code Quality:** genes.nf calls non-existent CLI commands.
genes.nf (line 63), genes.nf (line 80), genes.nf (line 97) call rnac genes fetch/preprocess/classify, but those are registered under rnac genes infer ... in genes.py (line 48). preprocess also unconditionally checks Path(regions_data) at genes.py (line 138), but the workflow passes no --regions_data.
Impact: the workflow fails before inference, or at preprocess after command names are fixed.
Options: Recommended: update workflow calls to rnac genes infer ... and fix/pass regions_data. Effort low; risk low; impact narrow; maintenance low. Alternative: add top-level CLI aliases. Effort low; risk medium due duplicate CLI surface; maintenance medium. Do nothing: pipeline remains broken.

**High, Architecture:** the upstream _flag is not actually used to gate work.
main.nf pipes precompute into genes at main.nf (line 48), but workflow genes ignores _flag and starts from Channel.of(true) / Channel.fromPath(...) at genes.nf (line 240).
Impact: genes can query/store against the DB before import/analyze/precompute has finished updating state.
Options: Recommended: thread _flag into the first real work channel so fetch_taxids and model fetches depend on upstream completion. Effort low-medium; risk low; impact local; maintenance low. Alternative: keep genes.nf standalone and remove it from main.nf until ordering is explicit. Effort low; risk medium operationally; maintenance medium. Do nothing: production race.

**High, Architecture/Data Safety:** metadata processing uses one shared DB table name while workflow can run per taxon in parallel.
process_metadata has no maxForks in genes.nf (line 198), and get_metadata drops/recreates global genes_urs_taxid at data.py (line 1075). Worker queries then join that same table.
Impact: concurrent taxa can overwrite each other’s lookup table, producing corrupt metadata or intermittent failures.
Options: Recommended: remove the shared table and pass URS lists directly via parameterized queries/CTEs. Effort medium; risk low; impact contained to metadata helpers; maintenance low. Alternative: generate unique table names per process and drop in finally. Effort medium; risk medium; maintenance medium. Do nothing: flaky metadata.

**High, Configuration:** params.genes.release is used but not defined in committed config.
Used at genes.nf (line 137) and genes.nf (line 157); config/genes.config (line 1) defines URLs and taxa_query, but no release.
Impact: documented standalone runs can pass null/missing release into Click integer options, failing or corrupting version fields.
Options: Recommended: add an explicit required release param with an early Nextflow validation error. Effort low; risk low; impact narrow; maintenance low. Alternative: derive release from DB in the workflow. Effort medium; risk medium; maintenance medium. Do nothing: hidden runtime failure.

**Medium, Edge Cases/Tests:** empty classification outputs are still not handled end-to-end.
classify can return an empty DataFrame at classify.py (line 299), but downstream init_genes, store_genes, and get_metadata assume name/members columns exist at data.py (line 489), data.py (line 532), and data.py (line 1058).
Impact: any taxon with no inferred multi-member genes can crash later stages.
Options: Recommended: branch/no-op empty taxa before init/merge/store/metadata, with typed empty outputs. Effort medium; risk low; impact local; maintenance low. Alternative: make every downstream helper accept empty/no-column JSON. Effort medium; risk medium; maintenance medium. Do nothing: sporadic taxon failures.

Tests
I ran git diff --check HEAD~5..HEAD; it passed. I did not run DB/Nextflow execution tests. There are no nf-test files, and the existing genes test coverage is still the legacy build path, not this new random-forest workflow.

**Review by Vicbeg — APPROVED** [review]
I think the changes look good. I have added a comment with some suggestions by Codex, in case you want to take a look at any of the potential issues it raises but I'll go ahead and approve.

**Comment by afg1** [comment]
Thanks! In response to codex's points:
1. The cli path is correct in the genes.nf, so I'm not sure where it got that from. Also, the check on genes.py line 138 is intentional, as the variable will be None if it isn't passed. I think this is fine
2. This comes from a misunderstanding of how the whole pipeline runs. We don't use main.nf so the _flag being used for synchronisation is irrelevant really. We _could_ use it, but the pipeline basically never runs end-to-end without incident
3. This might be a good point!
4. Release should be set in the local.config, which we don't commit
5. I think these are no-op be definition if the provided genes file is empty, so I don't think anything needs to be done

So, the data race identified in the metadata calculation is probably real, and I'll fix that now. Claude reckons we can do it without a DB access, which will make things faster anyway.
